ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/composable/WithShadow.kt

```diff
@@ -0,0 +1,266 @@
+
+import androidx.compose.foundation.background
+import androidx.compose.foundation.border
+import androidx.compose.foundation.layout.Arrangement
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.PaddingValues
+import androidx.compose.foundation.layout.calculateEndPadding
+import androidx.compose.foundation.layout.calculateStartPadding
+import androidx.compose.foundation.layout.offset
+import androidx.compose.foundation.layout.padding
+import androidx.compose.foundation.layout.requiredSize
+import androidx.compose.foundation.shape.CircleShape
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.material3.MaterialTheme
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.BlurredEdgeTreatment
+import androidx.compose.ui.draw.blur
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.graphics.RectangleShape
+import androidx.compose.ui.graphics.Shape
+import androidx.compose.ui.graphics.toArgb
+import androidx.compose.ui.layout.Layout
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.Constraints
+import androidx.compose.ui.unit.IntOffset
+import androidx.compose.ui.unit.dp
+import com.revenuecat.purchases.paywalls.components.properties.ColorInfo
+import com.revenuecat.purchases.paywalls.components.properties.ColorScheme
+import com.revenuecat.purchases.paywalls.components.properties.Shadow
+import com.revenuecat.purchases.ui.revenuecatui.components.modifier.background
+import com.revenuecat.purchases.ui.revenuecatui.components.properties.ShadowStyle
+import com.revenuecat.purchases.ui.revenuecatui.components.properties.toShadowStyle
+
+/**
+ * Adds a [shadow] in the provided [shape] to the [content].
+ *
+ * @param margin If your [content] has a margin, provide that here. Otherwise the transparent pixels of the margin will
+ * participate in the shadow.
+ */
+@Composable
+internal fun WithShadow(
+    shadow: ShadowStyle,
+    shape: Shape,
+    margin: PaddingValues = PaddingValues(all = 0.dp),
+    content: @Composable () -> Unit,
+) {
+    // We cannot use the built-in .shadow() modifier, as it takes very different parameters from what we have in our
+    // ShadowStyle. WithShadow also can't be a modifier itself, because the modifiers we use to draw the shadow
+    // (offset, blur and background) cannot be part of the same modifier chain as those applied to the content. If they
+    // are, they will be combined in ways we don't want.
+    // We might be able to achieve a similar result using a .drawBehind() modifier, but we would need to figure out how
+    // to draw gradients on a canvas.
+    Layout(
+        content = {
+            // Content
+            content()
```

Comment:
Just to make sure, will `Layout` put the `content` "on top" of the shadow?

Reply:
Yes, that happens below, where we first place the `shadowPlaceable` and then the `contentPlaceable`.